[addons] make binary addon management independent

AlwinEsch opened the pull request:
## Description
This changes add a new binary addon manager who handle the
addons on request.

This way allow the use of several instances on the same addon.

Currently only used on Visualization and Screensaver, but other binary
addon types becomes on next steps changed to use them here.

Further becomes on next changes everything related to binary addons
moved to the new folder "./xbmc/addons/binary-addons".

The old system is not changed!

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

tamland commented:
Yes, architecturally I think this is fine.

Ideally GetInstalledBinaryAddonPaths would return a CAddon with all the info CBinaryAddonManger needs instead of the paths directly. But if you can't make it work, that's ok. I will be easy to fix later once CAddonDll and CAddon are separated.

ksooo commented on xbmc/addons/AddonManager.h:
```diff
@@ -117,6 +117,8 @@ namespace ADDON
 
     bool GetInstallableAddons(VECADDONS& addons, const TYPE &type);
 
+    bool GetInstalledBinaryAddonPaths(std::vector<std::pair<std::string, bool>>& binaryAddonList);
```
Could you please add documentation for the new member function? Especially the content of the vector is not self explaining.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.cpp:
```diff
@@ -0,0 +1,34 @@
+CBinaryAddonBase::CBinaryAddonBase(const std::string& addonPath)
+{
+
+  m_usable = false;
```
blank line should be removed. can "m_usable" initialized in ctor init list?

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.cpp:
```diff
@@ -0,0 +1,34 @@
+CBinaryAddonBase::~CBinaryAddonBase()
+{
+
+}
```
No need for this. "~CBinaryAddonBase() = default;" in the header file is sufficient.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.h:
```diff
@@ -0,0 +1,48 @@
+    ~CBinaryAddonBase();
+
+    const std::string& ID() { return m_id; }
+    const std::string& Path() { return m_path; }
```
these two member functions should be "const".

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,58 @@
+CBinaryAddonManager::CBinaryAddonManager()
+{
+
+}
```
no need for an empty definition. "= default;" in the header does the trick.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,58 @@
+CBinaryAddonManager::~CBinaryAddonManager()
+{
+
+}
```
no need for an empty definition. "= default;" in the header does the trick.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,58 @@
+  std::vector<std::pair<std::string, bool>> binaryAddonList;
+  if (!CAddonMgr::GetInstance().GetInstalledBinaryAddonPaths(binaryAddonList))
+  {
+    CLog::Log(LOGERROR, "Binary Addons: failed to get list");
```
could you adjust the wording to follow the log "syntax" you're using everywhere else (\_\_FUNCTION\_\_ and the like)?

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,58 @@
+
+  for (auto addon : binaryAddonList)
+  {
+    BinaryAddonBase base = std::make_shared<CBinaryAddonBase>(addon.first);
```
just a question: why don't you fill the vector with the addon instances instaed of strings in<code>CAddonMgr::GetInstalledBinaryAddonPaths</code> as you're here instanciate them unconditionally? (member function should be renamed to GetInstalledBinaryAddons, then).

Something like comes to mind:
<code>
typedef std::vector<std::pair<std::shared_ptr<CBinaryAddonBase>, bool>> BinaryAddonList;
</code>
<code>
bool GetInstalledBinaryAddons(BinaryAddonList& binaryAddonList);
</code>

AlwinEsch replied:
A from CAddonManager created CAddon is not complete usable, it can then only read the Path from him (and maybe things like names ...) and then destroy it again.

But this becomes also done from the xml parser on CBinaryAddonBase.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,58 @@
+      continue;
+    m_installedAddons[base->ID()] = base;
+    if (addon.second)
+      m_enabledAddons[base->ID()] = base;
```
what's the advantage of having two vectors here, one for all usable installed addons, another for the usable installed enabled addons? The addons contained in the second vector are 100% subset of the addons in the first vector. would it be more efficient to have just one vector containing the addon along with its 'enabled' flag? => header change: <code>std::vector<std::pair<std::shared_ptr<CBinaryAddonBase>, bool>> m_usableAddons;</code>?

AlwinEsch replied:
Was from me separated to two places to have more efficent, to get wanted type.
If enabled becomes requested only the list from there must be checked, otherwise the whole complete with all, was come on my old way, where every addon was stored like them and mostly only few are enabled.

AlwinEsch commented:
The request is now complete updated and use a working way.

ksooo reviewed (changes requested):
some questions and suggestions...

AlwinEsch commented on xbmc/ServiceManager.cpp:
```diff
@@ -78,6 +79,13 @@ bool CServiceManager::Init2()
+  m_binaryAddonManager.reset(new ADDON::CBinaryAddonManager());
+  if (!m_binaryAddonManager->Init())
+  {
+    CLog::Log(LOGFATAL, "CServiceManager::Init: Unable to start CBinaryAddonManager");
```
Always this c&p ;)

ksooo commented on xbmc/addons/AddonManager.h:
```diff
@@ -117,6 +117,8 @@ namespace ADDON
 
     bool GetInstallableAddons(VECADDONS& addons, const TYPE &type);
 
+    bool GetInstalledBinaryAddonPaths(std::vector<std::tuple<std::string, bool, std::string>>& binaryAddonList);
```
+ please add documentation, esp. the content of the vector is far from being self-explaining. ;-)
+ you should introduce a typedef for the complex type used here:
<code>typedef std::vector<std::tuple<std::string, bool, std::string>> BinaryAddonList;</code>

ksooo commented on xbmc/addons/binary-addons/AddonDll.cpp:
```diff
@@ -47,18 +48,31 @@
-  m_interface = {0};
 }
 
+
```
no need for a second blank line.

ksooo commented on xbmc/addons/binary-addons/AddonDll.cpp:
```diff
@@ -385,6 +399,14 @@ void CAddonDll::DestroyInstance(const std::string& instanceID)
+  AddonPtr addon = CServiceBroker::GetBinaryAddonManager().GetRunningAddon(ID());
+  if (addon)
+    return addon;
+  return AddonPtr();
```
the whole implementation of this member function can be reduced to one line: <code>return CServiceBroker::GetBinaryAddonManager().GetRunningAddon(ID());</code> ;-)

AlwinEsch replied:
Ah yes, a remnant from my test of them :)

ksooo commented on xbmc/addons/binary-addons/AddonInstanceHandler.cpp:
```diff
@@ -19,35 +19,70 @@
+
+  m_addon = m_addonBase->GetAddon(this);
+  if (!m_addon)
+    CLog::Log(LOGFATAL, "ADDON::IAddonInstanceHandler: Tried to get add-on '%s' who not available!", m_addonBase->ID().c_str());
```
use \_\_FUNCTION\_\_ instead of hardcoded function name

ksooo commented on xbmc/addons/binary-addons/AddonInstanceHandler.cpp:
```diff
@@ -59,8 +94,7 @@ bool IAddonInstanceHandler::CreateInstance(KODI_HANDLE instance)
 {
-  if (m_addon)
-    m_addon->DestroyInstance(m_instanceId);
+  m_addon->DestroyInstance(m_instanceId);
```
is it guaranteed that m_addon is always set here? In all (many?) other functions you take care to check before access. why not here?

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.cpp:
```diff
@@ -0,0 +1,383 @@
+  CXBMCTinyXML xmlDoc;
+  if (!xmlDoc.LoadFile(addonXmlPath))
+  {
+    CLog::Log(LOGERROR, "CBinaryAddonBase: Unable to load '%s', Line %d\n%s",
```
la la la, singing... \_\_FUNCTION\_\_ ...

AlwinEsch replied:
:-D already done

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.cpp:
```diff
@@ -0,0 +1,383 @@
+  int contents = 0;
+  for (auto addonType : m_types)
+    contents += addonType.ProvidedSubContents();
+  return contents > 0 ? true : false;
```
could be just <code>return contents > 0;</code>

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.cpp:
```diff
@@ -0,0 +1,383 @@
+{
+  if (handler == nullptr)
+  {
+    CLog::Log(LOGERROR, "ADDONS: GetAddon for Id '%s' called with empty instance handler", ID().c_str());
```
la la la, singing... \_\_FUNCTION\_\_ ... here and everywhere else in this PR.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.cpp:
```diff
@@ -0,0 +1,383 @@
+
+  CSingleLock lock(m_critSection);
+
+  const auto& presentHandler = m_activeAddonHandlers.find(handler);
```
no need for const and & around auto

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.cpp:
```diff
@@ -0,0 +1,383 @@
+  {
+    m_activeAddon.reset();
+  }
+
```
this empty line could be removed.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonBase.cpp:
```diff
@@ -0,0 +1,383 @@
+  if (m_types.empty())
+  {
+    CBinaryAddonType addonType(ADDON_UNKNOWN, this, nullptr);
+    m_types.push_back(addonType);
```
no need for the variable here. could just read <code>m_types.emplace_back(CBinaryAddonType(ADDON_UNKNOWN, this, nullptr));</code>

ksooo commented on xbmc/addons/binary-addons/BinaryAddonExtensions.cpp:
```diff
@@ -0,0 +1,155 @@
+  if (id.empty())
+    return m_children;
+
+  EXT_ELEMENTS childs;
```
the plural of "child" is "children", not "childs".

ksooo commented on xbmc/addons/binary-addons/BinaryAddonExtensions.h:
```diff
@@ -0,0 +1,69 @@
+  struct SExtValue
+  {
+    SExtValue(const std::string& strValue) : str(strValue) { }
+    const std::string& asString()  const { return str; }
```
one of the two blanks between "ring()" and "const {" could be removed here.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonExtensions.h:
```diff
@@ -0,0 +1,69 @@
+  class CBinaryAddonExtensions;
+  typedef std::vector<std::pair<std::string, CBinaryAddonExtensions> > EXT_ELEMENTS;
+  typedef std::vector<std::pair<std::string, SExtValue> > EXT_VALUE;
+  typedef std::vector<std::pair<std::string, EXT_VALUE> > EXT_VALUES;
```
with c++11, no need for a blank between two <code>></code> anymore.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,240 @@
+{
+  CAddonMgr::GetInstance().Events().Subscribe(this, &CBinaryAddonManager::OnEvent);
+
+  std::vector<std::tuple<std::string, bool, std::string>> binaryAddonList;
```
"auto" instead of the spaghetti type name.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,240 @@
+
+  for (auto addon : binaryAddonList)
+  {
+    BinaryAddonBasePtr base = std::make_shared<CBinaryAddonBase>();
```
just a question: why don't you fill the vector with the addon instances instead of strings in <code>CAddonMgr::GetInstalledBinaryAddonPaths</code> as you're instanciate them unconditionally here? (member function should be renamed to GetInstalledBinaryAddons, then).

AlwinEsch replied:
A from CAddonManager created CAddon is not complete usable, it can then only read the Path from him (and maybe things like names ...) and then must it be destroyed (The CAddonDll need them also as parent).

But this becomes also done from the xml parser on CBinaryAddonBase.

ksooo replied:
okay, you are the boss, was just a question anyway

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,240 @@
+    }
+    m_installedAddons[base->ID()] = base;
+    if (std::get<1>(addon))
+      m_enabledAddons[base->ID()] = base;
```
what's the advantage of having two vectors here, one for all usable installed addons, another for the usable installed enabled addons? The addons contained in the second vector are 100% subset of the addons in the first vector. would it be more efficient to have just one vector containing the addon along with its 'enabled' flag? => header change: <code>BinaryAddonList m_usableAddons;</code>?

Or is there a real performance gain, because the functions using the two vectors are called really very often?

AlwinEsch replied:
Was from me separated to two places to have more efficent, to get wanted type.
If enabled becomes requested only the list from there must be checked, otherwise the whole complete with all, was come on my old way, where every addon was stored like them and mostly only few are enabled.

ksooo replied:
okay, if it was a well profen decision for performance reasons, I'm fine with this.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,240 @@
+  else
+    addons = &m_installedAddons;
+
+  for (auto& info : *addons)
```
no need for "&" after "auto"

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,240 @@
+{
+  CSingleLock lock(m_critSection);
+
+  const auto& addon = m_installedAddons.find(addonId);
```
no need for "const" and "&" before/after "auto" here and everywhere else in this PR

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,240 @@
+{
+  CSingleLock lock(m_critSection);
+
+  std::vector<std::tuple<std::string, bool, std::string>> binaryAddonList;
```
auto instead of full type name?

AlwinEsch replied:
Has testet now, there does not work a auto, I replace it with a typedef
```
  typedef std::tuple<std::string, bool, std::string> BINARY_ADDON_LIST_ENTRY;
  typedef std::vector<BINARY_ADDON_LIST_ENTRY> BINARY_ADDON_LIST;
```

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,240 @@
+      }
+      m_installedAddons[base->ID()] = base;
+      if (std::get<1>(addon))
+        m_enabledAddons[base->ID()] = base;
```
at the first glance, the above code lines look pretty much the same as found in <code>bool CBinaryAddonManager::Init()</code>. maybe you should factor out the identical code?

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.h:
```diff
@@ -0,0 +1,113 @@
+     * @param[in] type Add-on type to check installed
+     * @return true if given type is installed
+     */
+    bool HasInstalledAddons(const TYPE &type);
```
this member function could be const.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.h:
```diff
@@ -0,0 +1,113 @@
+     * @param[in] type Add-on type to check enabled
+     * @return true if given type is enabled
+     */
+    bool HasEnabledAddons(const TYPE &type);
```
this member function could be const.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.h:
```diff
@@ -0,0 +1,113 @@
+     *                        If a type id becomes added are only add-ons
+     *                        returned who match them. Default is for all types.
+     */
+    void GetAddonInfos(BinaryAddonBaseList& addonInfos, bool enabledOnly, const TYPE &type);
```
this member function could be const.

AlwinEsch replied:
Has tried with several ways, has produced always errors, with the try to write on addonInfos.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.h:
```diff
@@ -0,0 +1,113 @@
+     *                 not a nullptr is returned
+     * @return add-on information pointer of installed add-on
+     */
+    const BinaryAddonBasePtr GetInstalledAddonInfo(const std::string& addonId, const TYPE &type = ADDON_UNKNOWN);
```
this member function could be const.

ksooo reviewed (changes requested):
Here we go. Lots of stuff...

tamland commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,237 @@
+  CSingleLock lock(m_critSection);
+
+  BINARY_ADDON_LIST binaryAddonList;
+  CAddonMgr::GetInstance().GetInstalledBinaryAddonPaths(binaryAddonList);
```
@AlwinEsch Can you (as we discussed) change `GetInstalledBinaryAddonPaths` to in addition to the paths return `CAddon`s directly, then pass them on to`CBinaryAddonBase`? This way, parsing of all the common info doesn't have to be duplicated.

ksooo replied:
This is what I suggested as well.

AlwinEsch replied:
I change it

AlwinEsch committed:
[addons] update on CAddonInfo setters to set values

Needed from external places if something on CAddon construction need
to changed.

AlwinEsch committed:
[addons} add option to ignore extension parse on manager

Is for binary addons who do this job byself and not needed
from here.

AlwinEsch commented:
Is complete updated, instead of the CAddon use it CAddonInfo who was created by CAddonBuilder.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,231 @@
+  CSingleLock lock(m_critSection);
+
+  auto addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end() && addon->second)
```
Is it really possible that an empty addoninfo is stored in the map? I think the addon->second check can be removed.

ksooo commented on xbmc/addons/binary-addons/BinaryAddonType.cpp:
```diff
@@ -0,0 +1,101 @@
+  return URIUtils::AddFileToFolder(m_path, m_libname);
+}
+
+const char* CBinaryAddonType::GetPlatformLibraryName(const TiXmlElement* element)
```
This member function should be static or at least const.

ksooo reviewed (changes requested):
Very  nice. Just some nitpicking...

AlwinEsch commented on xbmc/ServiceManager.cpp:
```diff
@@ -71,13 +72,20 @@ bool CServiceManager::Init2()
 
+  if (!m_binaryAddonManager->Init())
+  {
+    CLog::Log(LOGFATAL, "CServiceManager::Init: Unable to initialize CBinaryAddonManager");
```
I change it also on the other parts on CServiceManager :)

AlwinEsch replied:
Was the last one :D, them on moved CAddonDll becomes updated on other request.

tamland commented:
> Is complete updated, instead of the CAddon use it CAddonInfo who was created by CAddonBuilder.

Why didn't using CAddon work?

AlwinEsch commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -564,6 +567,40 @@ bool CAddonMgr::FindInstallableById(const std::string& addonId, AddonPtr& result
   return true;
 }
 
+bool CAddonMgr::GetInstalledBinaryAddons(BINARY_ADDON_LIST& binaryAddonList)
```
Here it can not "const", becomes in big conflict on much places then

ksooo replied:
I'm sure it can. Try harder. Mostly it does not work, because functions caked from this method are not convert, but they could and should be. See for example my other comment re GetPlatformLibraryName which currently is not const, but could even be static.

ksooo replied:
but yeah if it is too much effort, leave it as it is for now. can be optimized later.

AlwinEsch commented:
Make no sense to use, it need only to have the CAddonInfo (who becomes also stored) to have for binary addon creation.

CAddon becomes created with CAddonInfo as value, the binary part need then only to read the CAddonInfo and destroy the CAddon.

With that he gets it directly:
```cpp
binaryAddonList.push_back(BINARY_ADDON_LIST_ENTRY(!IsAddonDisabled(cp_addon->identifier), std::move(builder.GetAddonInfo())));
```

Otherwise must be the
```cpp
AddonPtr addon;
...
builder.Build();` 
binaryAddonList.push_back(BINARY_ADDON_LIST_ENTRY(!IsAddonDisabled(cp_addon->identifier), addon));
...
```
added who then make never needed work.

AlwinEsch commented:
Another part is the performance, here the measured from me:

The first way where addon.xml was readed complete on CBinaryAddonBase
0.002619 seconds
0.002962 seconds
0.002603 seconds
0.002628 seconds
0.001032 seconds

Here with CAddonBuiler and CAddon
0.010468 seconds
0.010382 seconds
0.010420 seconds
0.010542 seconds
0.010504 seconds

Here the current on request
0.007633 seconds
0.007950 seconds
0.008571 seconds
0.007605 seconds
0.008558 seconds

Measured on a 8x "Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz" CPU with 38 binary addons.

On the first can be seen how big the work is where cpluff is only used to get path.

These are small times, but times on a larger addon quantity and a weaker system can give it greater differences.

AlwinEsch commented:
jenkins build this please

tamland commented:
> Make no sense to use, it need only to have the CAddonInfo (who becomes also stored) for binary addon creation.
>
> CAddon becomes created with CAddonInfo as value, the binary part need then only to read the CAddonInfo and destroy the CAddon.

Everything else that interface with addon manager use IAddon as the 'info' so it makes sense to do the same for binary addons. Even if you don't to directly *need* it. When you don't set an extension point there is 0 overhead of `Build()` (it just calls the CAddon constructor). If it's impossible to do now, at the very least be prepared to fix it later

tamland commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  for (auto addon : binaryAddonList)
+    AddAddonBaseEntry(addon);
```
Missing lock

AlwinEsch replied:
Becomes added :-) , had it omitted since this is called only at the Kodi start

tamland commented on xbmc/addons/binary-addons/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,231 @@
+bool CBinaryAddonManager::AddAddonBaseEntry(BINARY_ADDON_LIST_ENTRY& entry)
+{
+  BinaryAddonBasePtr base = std::make_shared<CBinaryAddonBase>(entry.second);
+  if (!base->Create())
```
Holding the internal lock while doing io and external calls like this is a recipe for deadlocks. I don't have an easy fix for this, but it's better to write initialization in such a way that it doesn't require calls to be made in-between reading and writing to m_installedAddons/m_enabledAddons

AlwinEsch committed:
[addons] make binary addon management independent

This changes add a new binary addon manager who handle the
addons on request.

This way allow the use of several instances on the same addon.

Currently only used on Visualization and Screensaver, but other binary
addon types becomes on next steps changed to use them here.

Further becomes on next changes everything related to binary addons
moved to the new folder "./xbmc/addons/binary-addons".

The old system is not changed!

AlwinEsch commented:
jenkins build this please
